### PR TITLE
[#10315] Refactor search result for instructor search

### LINKED ISSUE #10315: Services should not be accessing component related models
In `search.service`, there are many instances where the service transforms data into component related models. This breaks reusability of the services as it is tightly linked to the component.

**Task**
- Define new interfaces for the service to decouple them from component models
- Have the components accept these new interfaces and transform them to their respective models for use

### PR DESCRIPTION
**Pull request by moziliar**
Fixes #10315 

**Outline of Solution**

Created an interface `SearchResult` as a generic type for search functions that stores an `any` type `searchResult` field. All the raw search result data will be passed in, and will be processed in the module using the search result. 

The developer is to ensure that the type passed in matches the type needed in the other modules. (would defensive programming needed here? Type validation with exception when type mismatch happens?)

@t-cheepeng 
Would there be a need to wrap the rest of the result types into this as well? Else, the name should be less generic than `SearchResult`.

### CODE REVIEW

File: src/web/services/search.service.ts [hunk 1]
```diff
@@ -473,11 +459,10 @@ export class SearchService {
...
 
 /**
- * The typings for the response object returned by the instructor search service.
+ * Generic search result type.
  */
-export interface InstructorSearchResult {
-  searchStudentsTables: SearchStudentsListRowTable[];
-  searchCommentsTables: SearchCommentsTable[];
+export interface SearchResult {
+  searchResult: any;
```
Comment (xpdavid) [inline src/web/services/search.service.ts:465]: But right now we loss the type information by doing this?

File: src/web/services/search.service.ts [hunk 2]
```diff
@@ -473,11 +459,10 @@ export class SearchService {
 }
 
 /**
- * The typings for the response object returned by the instructor search service.
+ * Generic search result type.
  */
-export interface InstructorSearchResult {
-  searchStudentsTables: SearchStudentsListRowTable[];
-  searchCommentsTables: SearchCommentsTable[];
```
Comment (xpdavid) [inline src/web/services/search.service.ts:480]: Why can't this just be `comments: CommentSearchResult[]`?

File: src/web/services/search.service.ts [hunk 3]
```diff
@@ -473,11 +459,10 @@ export class SearchService {
 }
 
 /**
- * The typings for the response object returned by the instructor search service.
+ * Generic search result type.
  */
-export interface InstructorSearchResult {
-  searchStudentsTables: SearchStudentsListRowTable[];
```
Comment (xpdavid) [inline src/web/services/search.service.ts:479]: And similar here

**Comment by moziliar** [comment]
@xpdavid I see.... I must have misunderstood the intent behind this refactoring... Would change it then!

**Comment by t-cheepeng** [comment]
> Would there be a need to wrap the rest of the result types into this as well? 

You only need to refactor the InstructorSearchResult. The goal of the refactor is not to refactor the whole search service to return a common type. In fact, we want the type to be different as @xpdavid has mentioned.

**Comment by moziliar** [comment]
@t-cheepeng I have changed the refactoring to conform to the suggestions. Is this what you expect?

I exposed a search function for `InstructorPrivilege` too.

File: angular.json [hunk 4]
```diff
@@ -158,5 +158,8 @@
       }
     }
   },
-  "defaultProject": "teammates"
-}
+  "defaultProject": "teammates",
+  "cli": {
+    "analytics": false
+  }
+}
```
Comment (t-cheepeng) [inline angular.json:165]: unnecessary changes? u can disable angular analytics globally with  `ng analytics off` so you don't get these changes

File: src/web/app/pages-instructor/instructor-search-page/instructor-search-page.component.ts [hunk 5]
```diff
@@ -84,4 +106,85 @@ export class InstructorSearchPageComponent implements OnInit {
...
+    );
+    const coursesWithStudents: SearchStudentsListRowTable[] = distinctCourses.map(
+        (courseId: string) => ({
+          courseId,
+          students: Array.from(
+              new Set(
+                  students
+                      .filter((s: Student) => s.courseId === courseId),
+              ),
+          ).map((s: Student) => ({
```
Comment (t-cheepeng) [inline src/web/app/pages-instructor/instructor-search-page/instructor-search-page.component.ts:129]: 👍 . Yup this is what we were looking for with the refactoring!

**Review by t-cheepeng — CHANGES_REQUESTED** [review]

File: src/web/app/pages-instructor/instructor-search-page/instructor-search-page.component.ts (hunk 5 above)
Comment (moziliar) [reply src/web/app/pages-instructor/instructor-search-page/instructor-search-page.component.ts:129]: Got it. Could I just remove the tests for the functions removed? Or do I port them over to the components that are using them?

Comment (t-cheepeng) [reply src/web/app/pages-instructor/instructor-search-page/instructor-search-page.component.ts:129]: If you could port them over in this PR, it would be great

**Comment by moziliar** [comment]
@t-cheepeng migrated the relevant tests over.

**Comment by t-cheepeng** [comment]
> Now the test is showing some failure that I could not reproduce in local

Try and see the Travis logs. You should be able to find why your tests are failing. You have deleted something that is required

**Comment by moziliar** [comment]
> @t-cheepeng migrated the relevant tests over.
> Now the test is showing some failure that I could not reproduce in local :(

It's very strange that the import is still in my local clone..... I really have no idea why so:(

**Comment by wkurniawan07** [comment]
@moziliar hint: is your branch up-to-date with the latest `master`?

**Comment by moziliar** [comment]
> @moziliar hint: is your branch up-to-date with the latest `master`?

@wkurniawan07 yup, just rebased to the latest master, but got shot down again :(

**Review by t-cheepeng — APPROVED** [review]
LGTM

File: src/web/services/search.service.ts [hunk 6]
```diff
@@ -120,87 +116,10 @@ export class SearchService {
     return this.httpRequestService.get(ResourceEndpoints.SEARCH_COMMENTS, paramMap);
   }
 
-  getCoursesWithStudents(studentsRes: Students): SearchStudentsListRowTable[] {
-    const { students }: { students: Student[] } = studentsRes;
-
-    const distinctCourses: string[] = Array.from(
-      new Set(students.map((s: Student) => s.courseId)),
+  searchInstructorPrivilege(courseId: string, sectionName: string): Observable<InstructorPrivilege> {
+    return this.instructorService.loadInstructorPrivilege(
```
Comment (xpdavid) [inline src/web/services/search.service.ts:120]: You can just call the method directly. 

- There is no need to use search service as a proxy
- The naming is also a bit misleading as we are not searching instructor privilege.

**Review by xpdavid — APPROVED** [review]
